### PR TITLE
panlint: Add option to ignore missing inclusion of components/config

### PR DESCRIPTION
**Pull request by wpoely86**
Panlint will not issue errors about the missing include for any component in the whitelist.

### CODE REVIEW

**Comment by wpoely86** [comment]
@stdweird have a look

**Comment by ned21** [comment]
What's the use case for this?

**Comment by wpoely86** [comment]
I find the `Component X in use, but component config has not been included` errors not really helpful. Especially with metaconfig. It means including a lot of extra lines and it doesn't say anything about including the schema.

So far I've mostly ignored the message but now we have Jenkins that is not happy because of it. This allows a way to get around it.

**Review by ned21 — CHANGES_REQUESTED** [review]
"whitelist" is not inclusive language - can you please rename all instances of it in this code? Given that it's really about ignoring a specific test, perhaps --ignore-missing-component-config would be a better name?

Although I also have concerns about this being a special case of a general issue, and a more configuration option would make sense?

**Comment by jrha** [comment]
Additionally, can you add tests for whatever you implement please.

File: panc/src/main/scripts/panlint/panlint.py [hunk 1]
```diff
@@ -449,8 +449,11 @@ def lint_line(line, components_included, first_line=False, allow_mvn_templates=F
     return (diagnoses, messages, problem_count, first_line)
 
 
-def lint_file(filename, allow_mvn_templates=False):
+def lint_file(filename, allow_mvn_templates=False, whitelist_components=None):
```
Comment (jrha) [inline panc/src/main/scripts/panlint/panlint.py:452]:
```suggestion
def lint_file(filename, allow_mvn_templates=False, ignore_components=None):
```

File: panc/src/main/scripts/panlint/panlint.py [hunk 2]
```diff
@@ -449,8 +449,11 @@ def lint_line(line, components_included, first_line=False, allow_mvn_templates=F
...
     """Run lint checks against all lines of a file."""
+    if whitelist_components is None:
+        whitelist_components = []
```
Comment (jrha) [inline panc/src/main/scripts/panlint/panlint.py:455]:
```suggestion
    if ignore_components is None:
        ignore_components = []
```

File: panc/src/main/scripts/panlint/panlint.py [hunk 3]
```diff
@@ -469,6 +472,8 @@ def lint_file(filename, allow_mvn_templates=False):
 
     # Get list of all component configs included in template
     components_included = RE_COMPONENT_INCLUDE.findall(raw_text)
+    # add whitelisted components
+    components_included.extend(whitelist_components)
```
Comment (jrha) [inline panc/src/main/scripts/panlint/panlint.py:476]:
```suggestion
    # add ignored components
    components_included.extend(ignore_components)
```

File: panc/src/main/scripts/panlint/panlint.py [hunk 4]
```diff
@@ -495,6 +500,8 @@ def main():
     parser.add_argument('--allow_mvn_templates', action='store_true', help='Allow use of maven templates')
     parser.add_argument('--always_exit_success', action='store_true',
                         help='Always exit cleanly even if problems are found')
+    parser.add_argument('--whitelist-components', type=str,
```
Comment (jrha) [inline panc/src/main/scripts/panlint/panlint.py:503]:
```suggestion
    parser.add_argument('--ignore-components', type=str,
```

File: panc/src/main/scripts/panlint/panlint.py [hunk 5]
```diff
@@ -514,9 +521,13 @@ def main():
         print('No files were provided, not doing anything')
         return 0
 
+    whitelist_components = None
+    if args.whitelist_components:
+        whitelist_components = args.whitelist_components.split(',')
```
Comment (jrha) [inline panc/src/main/scripts/panlint/panlint.py:526]:
```suggestion
    ignore_components = None
    if args.ignore_components:
        ignore_components = args.ignore_components.split(',')
```

File: panc/src/main/scripts/panlint/panlint.py [hunk 6]
```diff
@@ -514,9 +521,13 @@ def main():
...
+
     for path in args.paths:
         for filename in glob(path):
-            file_reports, file_problems = lint_file(filename, args.allow_mvn_templates)
+            file_reports, file_problems = lint_file(filename, args.allow_mvn_templates, whitelist_components)
```
Comment (jrha) [inline panc/src/main/scripts/panlint/panlint.py:530]:
```suggestion
            file_reports, file_problems = lint_file(filename, args.allow_mvn_templates, ignore_components)
```

**Comment by ned21** [comment]
Looking good - can you please squash the commits and amend the commit message of 7c891c0bca4673daf00c596772b596e978d3a99c?

**Comment by wpoely86** [comment]
@ned21 good to go

**Review by ned21 — APPROVED** [review]

**Comment by ned21** [comment]
Approved. Do we have a plan to do a new release of panc to include the updated panlint? Or do we just recommend sites use it from HEAD?

**Comment by jrha** [comment]
Sometime after the next general release yes, but I'd say using it from HEAD is fine too.

**Pull request merged**